clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4460,11 +4460,96 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   if (IsIAMCU && types::isCXX(Input.getType()))
     D.Diag(diag::err_drv_clang_unsupported) << "C++ for IAMCU";
 
+  // Handle depscan.
+  if (JA.getKind() == Action::DepscanJobClass) {
+    CmdArgs.push_back("-cc1depscan");
+
+    const OptSpecifier DepScanOpts[] = {
+        options::OPT_fdepscan_EQ,
+        options::OPT_fdepscan_share_EQ,
+        options::OPT_fdepscan_share_parent,
+        options::OPT_fdepscan_share_parent_EQ,
+        options::OPT_fno_depscan_share,
+        options::OPT_fdepscan_share_stop_EQ,
+        options::OPT_fdepscan_daemon_EQ,
+        options::OPT_fdepscan_prefix_map_EQ,
+        options::OPT_fdepscan_prefix_map_sdk_EQ,
+        options::OPT_fdepscan_prefix_map_toolchain_EQ};
+
+    Args.AddAllArgs(CmdArgs, DepScanOpts);
+
+    assert(Output.isFilename() && "Depscan needs to have file output");
+    CmdArgs.push_back("-o");
+    CmdArgs.push_back(Output.getFilename());
+    CmdArgs.push_back("-cc1-args");
+  }
+
   // Invoke ourselves in -cc1 mode.
   //
   // FIXME: Implement custom jobs for internal actions.
   CmdArgs.push_back("-cc1");
 
+  // Handle full response file input.
+  if (Inputs.front().getType() == types::TY_CMD) {
+    assert(Inputs.size() == 1 && "Only one response file input");
+    auto &II = Inputs.front();
+    assert(II.isFilename() && "Should be response file");
+    SmallString<128> InputName("@");
+    InputName += II.getFilename();
+    CmdArgs.push_back(C.getArgs().MakeArgString(InputName));
+
+    // FIXME: We overwrite the output and the action kind argument. Maybe there
+    // are better way to do that. Not all output kind is handled here.
+    if (isa<AssembleJobAction>(JA)) {
```

Comment:
Could we factor the code that handles the action into a function so we don't need to duplicate it?

Reply:
I added a comment here. This is to avoid horrible merge conflicts which I already hit. That should absolutely be a separate function when we upstreaming.

Reply:
Can we just skip this? If the depscanning action gets the response file via `-fdepscan-output` (or `-o`) and then *also* receives the follow-up action's `-emit-llvm` (or whatever) and `-o` as part of `-cc1-args`, then it can include the `-o` as part of the response file.

To be clear, I'm thinking this command should be:
```
% clang -cc1 @file.rsp
```
so `-###` ends up with either:
```
% clang -cc1depscan -fdepscan-output file.rsp -fdepscan-args=... --cc1-args -emit-llvm -o file.bc
% clang -cc1 @file.rsp
```
or
```
% clang -cc1depscan -o file.rsp -fdepscan-args=... --cc1-args -emit-llvm -o file.bc
% clang -cc1 @file.rsp
```

Reply:
To do what you are suggesting, we need depscan to know the output type, which is not easy currently, because:
* The job in the front can't peek into the job later to see what they are doing
* The dependency scanning job currently has the output type set to be a response file
* The final output type is set after combining a series job into one and it would be hard to set the final output when the depscan phase is created.

I will take a look to see if we can put some into phase combine stage and maybe it can be done there.

Reply:
Maybe you can add something to the end of `ToolSelector::getTool()`, right after the call to `combineWithPreprocessor()`, called `mixWithScanning()` or something?

I agree that it's a bit awkward. Phase combining feels like it should be an optimization, not a requirement, but in this case it'd be a requirement.

I suppose another approach (totally different from this patch) would be to add an `-fdepscan` option to `-cc1`... and `cc1_main()` is what connects to the daemon, etc.; then the `-cc1` is canonicalized and a new `CompilerInvocation` created and we compile from there. That model aligns less well with integrating with libclang / build system / etc. though.